src/windows/list.cpp

```diff
@@ -0,0 +1,687 @@
+
+    device->device = std::string(WCharToUtf8(deviceDetailData->DevicePath));
+
+    hDevice = CreateFileW(
+      deviceDetailData->DevicePath, handleOpenFlags,
+      FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
+      OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+
+    if (hDevice == INVALID_HANDLE_VALUE) {
+      errorCode = GetLastError();
+      device->error = "Couldn't open handle to device: Error " +
+      std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    int32_t deviceNumber = GetDeviceNumber(hDevice);
+
+    if (deviceNumber == -1) {
+      device->error = "Couldn't get device number";
+      result = false;
+      break;
+    }
+
+    device->raw = "\\\\.\\PhysicalDrive" + std::to_string(deviceNumber);
+
+    GetMountpoints(deviceNumber, &device->mountpoints);
+
+    hPhysical = CreateFileA(
+      device->raw.c_str(), handleOpenFlags,
+      FILE_SHARE_READ | FILE_SHARE_WRITE, NULL,
+      OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+
+    if (hPhysical == INVALID_HANDLE_VALUE) {
+      errorCode = GetLastError();
+      device->error = "Couldn't open handle to physical device: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    if (!GetDeviceSize(hPhysical, device)) {
+      errorCode = GetLastError();
+      device->error = "Couldn't get disk geometry: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    if (!GetAdapterInfo(hPhysical, device)) {
+      errorCode = GetLastError();
+      device->error = "Couldn't get device adapter descriptor: Error " +
+        std::to_string(errorCode);
+      result = false;
+      break;
+    }
+
+    // NOTE: No need to fail over this one,
+    // as we can safely default to a 512B block size
+    if (!GetDeviceBlockSize(hPhysical, device)) {
+      errorCode = GetLastError();
+      printf("[INFO] Couldn't get block size: Error %u\n", errorCode);
+    }
+
+    BOOL isWritable = DeviceIoControl(
+      hPhysical, IOCTL_DISK_IS_WRITABLE, NULL, 0,
+      NULL, 0, &size, NULL);
+
+    device->isReadOnly = !isWritable;
+  }
+
+  if (hDevice != INVALID_HANDLE_VALUE) {
+    CloseHandle(hDevice);
+  }
+
+  if (hPhysical != INVALID_HANDLE_VALUE) {
+    CloseHandle(hPhysical);
+  }
+
+  free(deviceDetailData);
+
+  return result;
+}
+
+std::vector<DeviceDescriptor> ListStorageDevices() {
+  HDEVINFO hDeviceInfo = NULL;
+  SP_DEVINFO_DATA deviceInfoData;
+  std::vector<DeviceDescriptor> deviceList;
+
+  DWORD i;
+  char *enumeratorName;
+  DeviceDescriptor device;
+
+  hDeviceInfo = SetupDiGetClassDevsA(
+    &GUID_DEVICE_INTERFACE_DISK, NULL, NULL,
+    DIGCF_PRESENT | DIGCF_DEVICEINTERFACE);
+
+  if (hDeviceInfo == INVALID_HANDLE_VALUE) {
+    printf("[ERROR] Invalid DeviceInfo handle\n");
+    return deviceList;
+  }
+
+  deviceInfoData.cbSize = sizeof(SP_DEVINFO_DATA);
+
+  for (i = 0; SetupDiEnumDeviceInfo(hDeviceInfo, i, &deviceInfoData); i++) {
```

Comment:
Looks a bit weird having this function-call as the middle-parameter of the for-statement?  (but maybe that's just because it's a programming style I'm not used to)